[c10d] Add dist.init()

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #162571
* #162549
* #163058
* #162770
* #162545
* __->__ #162529

Mainly to support a case where user don't want to call `init_process_group` which instantiates a real group that may not be used.

What `dist.init()` does:
```
    Initialize a distributed environment for current job. Once this API is
    called, `get_world_size()` is guaranteed to return the number of processes
    in current job, and `get_rank()` is guaranteed to return the rank of current
    process in the job.
```

With the proposed API, user can get group of interest via, e.g. DeviceMesh creation:
```
dist.init()
mesh = dist.init_device_mesh(...)
group = mesh.get_dim(...)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1526,6 +1551,83 @@ def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) ->
 
+@_exception_logger
+@_time_logger
+def init(
```
Should this be renamed `init_metadata` or something? Just `init()` implies it does everything rather than a restricted subset of initialization.

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1526,6 +1551,83 @@ def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) ->
+            world_size,
+            # TODO: should we support configurable timeout for rendezvous?
+        )
+        store, rank, world_size = next(rendezvous_iterator)
```
this is interesting -- we may not actually want to create a global store for performance reasons. tcpstore is fairly cheap but still hmm

we don't actually need it for get_rank()/get_world_size()

fduwjj replied:
Can we pass in the store so folks can use other stuff for example zues for rendezvous? Not sure...

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1526,6 +1551,83 @@ def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) ->
+    if store is None:
+        # Create one from environment variables
+        rendezvous_iterator = rendezvous(
+            "env://",
```
should we be consuming init_method to match init_process_group?

ezyang commented:
What's going on with the test failures?

ezyang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1526,6 +1551,83 @@ def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) ->
+    Initialize a distributed environment for current job. Once this API is
+    called, `get_world_size()` is guaranteed to return the number of processes
+    in current job, and `get_rank()` is guaranteed to return the rank of current
+    process in the job.
```
Can the doc explain why you want to do this? (Namely, to skip PG initialization). Can we describe exactly how many comms this initialization will take?

kwen2501 replied:
Thanks will add.

kwen2501 commented:
> What's going on with the test failures?

@ezyang Mostly from (the need for) backward support of file-based rendezvous -- which I tend to think "nah.." when coding the PR. But should be easily fixable.

kwen2501 commented on .lintrunner.toml:
```diff
@@ -140,6 +140,8 @@ include_patterns = [
 exclude_patterns = [
     '**/fb/**',
+    # TODO: Remove this once we clean up the code
+    'torch/distributed/distributed_c10d.py',
```
Yeah, it comes from this change:
```
-  _world = _World()
+  _world: Optional[_World] = None
```
Thanks for suggesting the global suppression.

kwen2501 commented:
@wconstab thanks for looking at it. The PR is just a sugarcoat of your "counter proposal".
Its little additional is that functions like `dist.get_rank()` lets user call it at any depth of the program stack.

wconstab commented:
Yea it's just a trade off, convenience vs "not having any global state". I'm ok either way

kwen2501 committed:
Update on "[c10d] Add dist.init()"


Mainly to support a case where user don't want to call `init_process_group` which instantiates a real group that may not be used.

What `dist.init()` does:
```
    Initialize a distributed environment for current job. Once this API is
    called, `get_world_size()` is guaranteed to return the number of processes
    in current job, and `get_rank()` is guaranteed to return the rank of current
    process in the job.
```

With the proposed API, user can get group of interest via, e.g. DeviceMesh creation:
```
dist.init()
mesh = dist.init_device_mesh(...)
group = mesh.get_dim(...)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]
